java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/protocol/WorkspaceServiceImpl.java

```diff
@@ -136,6 +138,81 @@
+                String uri = ((JsonPrimitive) params.getArguments().get(0)).getAsString();
+                String type = params.getArguments().size() > 1 ? ((JsonPrimitive) params.getArguments().get(1)).getAsString() : JavaProjectConstants.SOURCES_TYPE_JAVA;
+                FileObject file;
+                try {
+                    file = URLMapper.findFileObject(new URL(uri));
+                } catch (MalformedURLException ex) {
+                    Exceptions.printStackTrace(ex);
+                    return CompletableFuture.completedFuture(Collections.emptyList());
+                }
+                if (file == null) {
+                    return CompletableFuture.completedFuture(Collections.emptyList());
+                }
+                return server.asyncOpenFileOwner(file).thenApply(project -> {
+                    List<String> roots = new ArrayList<>();
+                    if (project != null) {
+                        for(SourceGroup sourceGroup : ProjectUtils.getSources(project).getSourceGroups(type)) {
+                            roots.add(Utils.toUri(sourceGroup.getRootFolder()));
+                        }
+                    }
+                    return roots;
+                });
+            }
+            case Server.JAVA_GET_PROJECT_CLASSPATH: {
+                String uri = ((JsonPrimitive) params.getArguments().get(0)).getAsString();
+                ClasspathInfo.PathKind kind = params.getArguments().size() > 1 ? ClasspathInfo.PathKind.valueOf(((JsonPrimitive) params.getArguments().get(1)).getAsString()) : ClasspathInfo.PathKind.COMPILE;
+                FileObject file;
+                try {
+                    file = URLMapper.findFileObject(new URL(uri));
+                } catch (MalformedURLException ex) {
+                    Exceptions.printStackTrace(ex);
+                    return CompletableFuture.completedFuture(Collections.emptyList());
+                }
+                if (file == null) {
+                    return CompletableFuture.completedFuture(Collections.emptyList());
+                }
+                return server.asyncOpenFileOwner(file).thenApply(project -> {
+                    HashSet<FileObject> roots = new HashSet<>();
+                    if (project != null) {
+                        for(SourceGroup sourceGroup : ProjectUtils.getSources(project).getSourceGroups(JavaProjectConstants.SOURCES_TYPE_JAVA)) {
+                            for (FileObject root : ClasspathInfo.create(sourceGroup.getRootFolder()).getClassPath(kind).getRoots()) {
+                                roots.add(root);
+                            }
+                        }
+                    }
+                    return roots.stream().map(fo -> Utils.toUri(fo)).collect(Collectors.toList());
+                });
+            }
+            case Server.JAVA_GET_PROJECT_PACKAGES: {
+                String uri = ((JsonPrimitive) params.getArguments().get(0)).getAsString();
```

Comment:
Nitpick: I have a vague feeling that all three code blocks could be factored out to a common method like
```
CompletableFuture<List<?>> collectFromGroups(String uri, String groupType, Function<SourceGroup, Object>)
```

Reply:
Common code factored out to `CompletableFuture<List<FileObject>> getSourceRoots(String uri, String type)`.